File: getting-started/telnyx-bot-signup/scripts/pow_solver.c
/*
 * pow_solver.c — Proof of Work solver for Telnyx bot signup challenge.
 *
 * Supported algorithms
 * --------------------
 *  sha256  : Uses OpenSSL SHA-256.  No challenge_config params needed.
 *  scrypt  : Memory-hard hash.  Uses OpenSSL EVP_PBE_scrypt (requires
 *            OpenSSL 1.1+ or LibreSSL 3.3+).  Reads n, r, p from
 *            challenge_config JSON.
 *
 * macOS / LibreSSL note
 * ---------------------
 *  The OpenSSL shipped with macOS is actually LibreSSL, which often lacks
 *  EVP_PBE_scrypt.  For scrypt support on macOS either:
 *    brew install openssl   (then compile with OPENSSL_PREFIX=/opt/homebrew/opt/openssl)
 *  or run this solver on Linux.  SHA-256 works fine with LibreSSL.
 *
 * Build
 * -----
 *  See the accompanying Makefile.  Quick start:
 *    make
 *    ./pow_solver <nonce> <leading_zero_bits> [algorithm] [challenge_config_json]
 *
 * Examples
 * --------
 *  # SHA-256 (default)
 *  ./pow_solver "abc123..." 22
 *
 *  # scrypt
 *  ./pow_solver "abc123..." 16 scrypt '{"n":4096,"r":8,"p":1}'
 *
 * Performance vs Python
 * ---------------------
 *  The C solver is typically 10–50x faster than the Python equivalent,
 *  depending on algorithm, hardware, and difficulty level.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <inttypes.h>

#include <openssl/sha.h>
#include <openssl/evp.h>

/* ------------------------------------------------------------------ helpers */

static int count_leading_zero_bits(const unsigned char *digest, size_t len) {
    int count = 0;
    for (size_t i = 0; i < len; i++) {
        unsigned char byte = digest[i];
        if (byte == 0) {
            count += 8;
        } else {
            /* count leading zero bits in this byte */
            for (int bit = 7; bit >= 0; bit--) {
                if ((byte >> bit) & 1) return count;
                count++;
            }
            return count;
        }
    }
    return count;
}

/* ------------------------------------------------------------ sha-256 solver */

static int64_t solve_sha256(const char *nonce, int leading_zero_bits) {
    char buf[4096];
    unsigned char digest[SHA256_DIGEST_LENGTH];

    for (int64_t i = 0; i < INT64_MAX; i++) {
        if (i > 0 && i % 1000 == 0) {
            fputc('.', stderr);
            fflush(stderr);
        }

        int len = snprintf(buf, sizeof(buf), "%s%" PRId64, nonce, i);
        if (len < 0 || (size_t)len >= sizeof(buf)) {
            fprintf(stderr, "\nERROR: nonce+counter string too long\n");
            return -1;
        }

        SHA256((unsigned char *)buf, (size_t)len, digest);

        if (count_leading_zero_bits(digest, SHA256_DIGEST_LENGTH) >= leading_zero_bits) {
            if (i >= 1000) fputc('\n', stderr);
            return i;
        }
    }

    fprintf(stderr, "\nERROR: No solution found within search space\n");
    return -1;
}

/* ------------------------------------------------------------- scrypt solver */

static int64_t solve_scrypt(const char *nonce, int leading_zero_bits,
                             uint64_t N, uint64_t r, uint64_t p) {
#if defined(OPENSSL_NO_SCRYPT) || !defined(EVP_PBE_SCRYPT_DEFINED)
    (void)nonce; (void)leading_zero_bits; (void)N; (void)r; (void)p;
    fprintf(stderr,
        "\nERROR: scrypt support was not compiled in.\n"
        "This build was compiled against LibreSSL or an OpenSSL version without scrypt.\n"
        "To use scrypt on macOS:\n"
        "    brew install openssl\n"
        "    make OPENSSL_PREFIX=/opt/homebrew/opt/openssl\n"
        "Or run the Python solver instead:\n"
        "    python3 pow_solver.py <nonce> <bits> scrypt '<config_json>'\n");
    return -1;
#else
    char buf[4096];
    unsigned char out[32];

    fprintf(stderr, "[scrypt] N=%" PRIu64 " r=%" PRIu64 " p=%" PRIu64
            " — solving, please wait...\n", N, r, p);

    for (int64_t i = 0; i < INT64_MAX; i++) {
        if (i > 0 && i % 1000 == 0) {
            fputc('.', stderr);
            fflush(stderr);
        }

        int len = snprintf(buf, sizeof(buf), "%s%" PRId64, nonce, i);
        if (len < 0 || (size_t)len >= sizeof(buf)) {
            fprintf(stderr, "\nERROR: nonce+counter string too long\n");
            return -1;
        }

        /* password = nonce+i, salt = nonce */
        if (EVP_PBE_scrypt(buf, (size_t)len,
                           (unsigned char *)nonce, strlen(nonce),
                           N, r, p, 0,
                           out, sizeof(out)) != 1) {
            fprintf(stderr, "\nERROR: EVP_PBE_scrypt failed\n");
            return -1;
        }

        if (count_leading_zero_bits(out, sizeof(out)) >= leading_zero_bits) {
            if (i >= 1000) fputc('\n', stderr);
            return i;
        }
    }

    fprintf(stderr, "\nERROR: No solution found within search space\n");
    return -1;
#endif
}

/* ----------------------------------------------------------------- JSON mini-parser
 * Extracts a single integer value from a flat JSON object like
 * {"n":4096,"r":8,"p":1} without pulling in a JSON library.
 */
static long json_get_long(const char *json, const char *key, long default_val) {
    /* look for "key": or "key" : */
    char search[64];
    snprintf(search, sizeof(search), "\"%s\"", key);
    const char *pos = strstr(json, search);
    if (!pos) return default_val;
    pos += strlen(search);
    while (*pos == ' ' || *pos == '\t' || *pos == ':') pos++;
    if (*pos == '\0') return default_val;
    return strtol(pos, NULL, 10);
}

/* ---------------------------------------------------------------------- main */

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr,
            "Usage: %s <nonce> <leading_zero_bits> [algorithm] [challenge_config_json]\n"
            "\n"
            "Examples:\n"
            "  # SHA-256 (default)\n"
            "  %s \"abc123...\" 22\n"
            "\n"
            "  # scrypt\n"
            "  %s \"abc123...\" 16 scrypt '{\"n\":4096,\"r\":8,\"p\":1}'\n",
            argv[0], argv[0], argv[0]);
        return 1;
    }

    const char *nonce            = argv[1];
    int         leading_zero_bits = atoi(argv[2]);
    const char *algorithm        = (argc > 3) ? argv[3] : "sha256";
    const char *config_json      = (argc > 4) ? argv[4] : "{}";

    int64_t solution = -1;

    if (strcmp(algorithm, "sha256") == 0) {
        solution = solve_sha256(nonce, leading_zero_bits);
    } else if (strcmp(algorithm, "scrypt") == 0) {
        uint64_t N = (uint64_t)json_get_long(config_json, "n", 4096);
        uint64_t r = (uint64_t)json_get_long(config_json, "r", 8);
        uint64_t p = (uint64_t)json_get_long(config_json, "p", 1);
        solution = solve_scrypt(nonce, leading_zero_bits, N, r, p);
    } else {
        fprintf(stderr,
            "ERROR: Unknown algorithm: %s\n"
            "Supported: sha256, scrypt\n"
            "If the server returned a new algorithm, use the Python solver as a\n"
            "reference — the loop structure is the same.\n", algorithm);
        return 1;
    }

    if (solution < 0) return 1;

    printf("%" PRId64 "\n", solution);
    return 0;
}

File: getting-started/telnyx-bot-signup/scripts/pow_solver.py
#!/usr/bin/env python3
"""
Proof of Work solver for Telnyx bot signup challenge.

Supported algorithms
--------------------
- sha256  : Uses hashlib.sha256. No challenge_config params needed.
- scrypt  : Memory-hard. Uses hashlib.scrypt (requires OpenSSL-linked
            Python — see note below). Reads N, r, p from challenge_config.

Extensibility
-------------
If the server returns an `algorithm` value not listed above, use the SHA-256
and scrypt implementations below as a reference. The protocol is always the
same brute-force loop — only the hash primitive changes. Use the algorithm
name and `challenge_config` keys to understand which hash function to call
and how to parameterize it.

macOS note on scrypt
--------------------
macOS ships Python linked against LibreSSL, which does NOT expose
hashlib.scrypt. If you see an AttributeError on `hashlib.scrypt`, you need
an OpenSSL-linked Python:

    brew install python   # then use /opt/homebrew/bin/python3
    # — or —
    run the solver on Linux

CLI usage
---------
    python3 pow_solver.py <nonce> <leading_zero_bits> [algorithm] [challenge_config_json]

Examples:
    # SHA-256 (default)
    python3 pow_solver.py "abc123..." 22

    # scrypt
    python3 pow_solver.py "abc123..." 16 scrypt '{"n": 4096, "r": 8, "p": 1}'
"""

import hashlib
import json
import sys


def _count_leading_zero_bits(digest_hex: str) -> int:
    """Return the number of leading zero bits in a hex digest string."""
    digest_int = int(digest_hex, 16)
    total_bits = len(digest_hex) * 4  # each hex char = 4 bits
    if digest_int == 0:
        return total_bits
    return total_bits - digest_int.bit_length()


def solve_sha256(nonce: str, leading_zero_bits: int) -> int:
    """Solve PoW using SHA-256."""
    for i in range(0, 2**63):
        if i > 0 and i % 1000 == 0:
            print(".", end="", flush=True, file=sys.stderr)
        digest = hashlib.sha256(f"{nonce}{i}".encode()).hexdigest()
        if _count_leading_zero_bits(digest) >= leading_zero_bits:
            if i >= 1000:
                print(file=sys.stderr)  # newline after progress dots
            return i
    raise RuntimeError("No solution found within search space")


def solve_scrypt(nonce: str, leading_zero_bits: int, config: dict) -> int:
    """
    Solve PoW using scrypt.

    Parameters are read from challenge_config:
      - n  : CPU/memory cost factor (e.g. 4096)
      - r  : block size (e.g. 8)
      - p  : parallelisation factor (e.g. 1)

    The salt is the nonce string encoded as UTF-8 bytes.
    The password is (nonce + str(i)) encoded as UTF-8 bytes.
    """
    if not hasattr(hashlib, "scrypt"):
        print(
            "\nERROR: hashlib.scrypt is not available on this Python installation.\n"
            "macOS ships Python linked against LibreSSL, which does not expose scrypt.\n"
            "To fix this, use an OpenSSL-linked Python:\n"
            "    brew install python   # then use /opt/homebrew/bin/python3\n"
            "    — or — run this solver on Linux.",
            file=sys.stderr,
        )
        sys.exit(1)

    n = config.get("n", 4096)
    r = config.get("r", 8)
    p = config.get("p", 1)

    # The salt is the nonce string itself as UTF-8 bytes (not hex-decoded).
    salt = nonce.encode()

    print(
        f"[scrypt] N={n} r={r} p={p} — solving, please wait...",
        file=sys.stderr,
    )

    for i in range(0, 2**63):
        if i > 0 and i % 1000 == 0:
            print(".", end="", flush=True, file=sys.stderr)
        password = f"{nonce}{i}".encode()
        raw = hashlib.scrypt(password, salt=salt, n=n, r=r, p=p, dklen=32)
        digest_hex = raw.hex()
        if _count_leading_zero_bits(digest_hex) >= leading_zero_bits:
            if i >= 1000:
                print(file=sys.stderr)  # newline after progress dots
            return i
    raise RuntimeError("No solution found within search space")


def solve(
    nonce: str,
    leading_zero_bits: int,
    algorithm: str = "sha256",
    config: dict = None,
) -> int:
    """Dispatch to the correct solver based on algorithm name."""
    if config is None:
        config = {}

    if algorithm == "sha256":
        return solve_sha256(nonce, leading_zero_bits)
    elif algorithm == "scrypt":
        return solve_scrypt(nonce, leading_zero_bits, config)
    else:
        raise ValueError(
            f"Unknown algorithm: {algorithm!r}. "
            "Supported: sha256, scrypt. "
            "If the server returned a new algorithm, use the sha256 and scrypt "
            "implementations as a reference — the loop structure is always the same."
        )


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print(
            f"Usage: {sys.argv[0]} <nonce> <leading_zero_bits> [algorithm] [challenge_config_json]",
            file=sys.stderr,
        )
        sys.exit(1)

    nonce = sys.argv[1]
    leading_zero_bits = int(sys.argv[2])
    algorithm = sys.argv[3] if len(sys.argv) > 3 else "sha256"
    config = json.loads(sys.argv[4]) if len(sys.argv) > 4 else {}

    solution = solve(nonce, leading_zero_bits, algorithm, config)
    print(solution)
